### PR TITLE
Make work

### PR DESCRIPTION
**Pull request by julialongtin**

### CODE REVIEW

File: tasks/failover.yml [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    name: redis-sentinel
+    state: started
+    daemon_reload: true
+
+- name: reconfigure sentinels to forget their current master
+  shell: redis-cli -p {{ redis_sentinel_port }} sentinel remove mymaster
+  when: ansible_hostname != 'instance'
+
+- name: set redis to run outside of protected mode
+  shell: redis-cli config set protected-mode no
```
[the remark this answers is not in the document]
Comment (julialongtin) [reply tasks/failover.yml]: the rewrite does not use the redis_cluster.conf.

File: tasks/failover.yml [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  
+- name: restart redis on slaves.
+  systemd:
+    name: redis
+    state: restarted
+    daemon_reload: true
+  when: ansible_hostname != 'instance'
+
+- name: add slave nodes as slave of master.
+  shell: MASTER=$(ip addr | grep inet | grep eth0 | sed "s=.*inet \(.*\)/.*=\1=") && {{ groups['redis_slave'] | map('extract', hostvars, ['ansible_eth0', 'ipv4', 'address']) | map('regex_replace', '^(.*)$','redis-cli -h \1 slaveof $MASTER 6379') | join (' && ') }}
```
[the remark this answers is not in the document]
Comment (julialongtin) [reply tasks/failover.yml]: done.

File: tasks/failover.yml [hunk 3]
```diff
@@ -0,0 +1,53 @@
+---
+- name: pause sentinels
+  systemd:
+    name: redis-sentinel
+    state: stopped
+    daemon_reload: true
+
+- name: force sentinels to announce an IP.
+  shell: grep -qE -- "^sentinel announce-ip" "/etc/redis/sentinel.conf" || echo "sentinel announce-ip "$(ip addr | grep inet | grep eth0 | sed "s=.*inet \(.*\)/.*=\1=") >> /etc/redis/sentinel.conf
```
[the remark this answers is not in the document]
Comment (julialongtin) [reply tasks/failover.yml]: done.

File: tasks/failover.yml [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+- name: set redis to run outside of protected mode
+  shell: redis-cli config set protected-mode no
+  
+- name: restart redis on slaves.
+  systemd:
+    name: redis
+    state: restarted
+    daemon_reload: true
+  when: ansible_hostname != 'instance'
```
[the remark this answers is not in the document]
Comment (julialongtin) [reply tasks/failover.yml]: Did something similar.

File: tasks/cluster.yml [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    state: restarted
+    daemon_reload: true
+
+- name: meet cluster nodes
+  shell: "{{ groups['redis'] | map('extract', hostvars, ['ansible_eth0', 'ipv4', 'address']) | map('regex_replace', '^(.*)$','redis-cli cluster meet \\1 6379') | join (' && ') }}"
+  when: ansible_hostname == 'instance'
+
+- name: add slots to primary node
+  shell: redis-cli cluster addslots $(seq 0 16383)
+  when: ansible_hostname == 'instance'
```
[the remark this answers is not in the document]
Comment (julialongtin) [reply tasks/cluster.yml]: did not use that in this rewrite.

**Comment by julialongtin** [comment]
for your review.

**Comment by julialongtin** [comment]
When i shut down the redis master process (/etc/init.d/redis-server stop), I see a failover. I'll start doing progressively worse actions to it, in order to reproduce the failure you're seeing.

**Comment by julialongtin** [comment]
so, running 'poweroff' on a node that is master results in a correct failover, and kill -9'ing the redis-server that is master just causes it to restart... and with one node powered off, shutting off the redis process on the master (service redis-server stop) just results in the last redis server becoming master. tabling this until i see you again, so i can see how you're producing a failure mode.

**Comment by jschaul** [comment]
```
apt-cache show redis-server | grep -i version
Version: 5:4.0.9-1ubuntu0.1
```

sentinel logs showing failover not happening:

```
40:X 03 May 12:32:56.269 # Sentinel ID is 71c9288565806da082b428463d07735bd08a9427
40:X 03 May 12:32:56.269 # +monitor master mymaster 172.17.0.4 6379 quorum 2
40:X 03 May 12:33:26.276 # +sdown slave 172.17.0.1:6379 172.17.0.1 6379 @ mymaster 172.17.0.4 6379
40:X 03 May 13:29:51.254 # -monitor master mymaster 172.17.0.4 6379
40:X 03 May 13:29:56.861 # +monitor master mymaster 172.17.0.4 6379 quorum 2
40:X 03 May 13:29:56.902 * +slave slave 172.17.0.1:6379 172.17.0.1 6379 @ mymaster 172.17.0.4 6379
40:X 03 May 13:29:58.817 * +sentinel sentinel 5549d29a0710c611c7ea55c95d51085ca5d8913c 172.17.0.2 26379 @ mymaster 172.17.0.4 6379
40:X 03 May 13:29:59.147 * +sentinel sentinel aa7f6fdf6a9b24b879374e78485dd9e0462d9f70 172.17.0.4 26379 @ mymaster 172.17.0.4 6379
40:X 03 May 13:30:01.886 # +reset-master master mymaster 172.17.0.4 6379
40:X 03 May 13:30:02.879 * +sentinel sentinel 5549d29a0710c611c7ea55c95d51085ca5d8913c 172.17.0.2 26379 @ mymaster 172.17.0.4 6379
40:X 03 May 13:30:02.909 * +sentinel sentinel aa7f6fdf6a9b24b879374e78485dd9e0462d9f70 172.17.0.4 26379 @ mymaster 172.17.0.4 6379
40:X 03 May 13:30:07.034 * +slave slave 172.17.0.1:6379 172.17.0.1 6379 @ mymaster 172.17.0.4 6379
40:X 03 May 13:30:37.087 # +sdown slave 172.17.0.1:6379 172.17.0.1 6379 @ mymaster 172.17.0.4 6379
40:X 03 May 13:31:14.183 # +sdown master mymaster 172.17.0.4 6379
40:X 03 May 13:31:14.184 # +sdown sentinel aa7f6fdf6a9b24b879374e78485dd9e0462d9f70 172.17.0.4 26379 @ mymaster 172.17.0.4 6379
40:X 03 May 13:31:14.250 # +odown master mymaster 172.17.0.4 6379 #quorum 2/2
40:X 03 May 13:31:14.250 # +new-epoch 2
40:X 03 May 13:31:14.250 # +try-failover master mymaster 172.17.0.4 6379
40:X 03 May 13:31:14.257 # +vote-for-leader 71c9288565806da082b428463d07735bd08a9427 2
40:X 03 May 13:31:14.270 # 5549d29a0710c611c7ea55c95d51085ca5d8913c voted for 71c9288565806da082b428463d07735bd08a9427 2
40:X 03 May 13:31:14.329 # +elected-leader master mymaster 172.17.0.4 6379
40:X 03 May 13:31:14.329 # +failover-state-select-slave master mymaster 172.17.0.4 6379
40:X 03 May 13:31:14.388 # -failover-abort-no-good-slave master mymaster 172.17.0.4 6379
40:X 03 May 13:31:14.465 # Next failover delay: I will not start a failover before Fri May  3 13:37:14 2019
40:X 03 May 13:37:14.694 # +new-epoch 3
40:X 03 May 13:37:14.695 # +try-failover master mymaster 172.17.0.4 6379
40:X 03 May 13:37:14.709 # +vote-for-leader 71c9288565806da082b428463d07735bd08a9427 3
40:X 03 May 13:37:14.727 # 5549d29a0710c611c7ea55c95d51085ca5d8913c voted for 71c9288565806da082b428463d07735bd08a9427 3
40:X 03 May 13:37:14.761 # +elected-leader master mymaster 172.17.0.4 6379
40:X 03 May 13:37:14.761 # +failover-state-select-slave master mymaster 172.17.0.4 6379
40:X 03 May 13:37:14.838 # -failover-abort-no-good-slave master mymaster 172.17.0.4 6379
40:X 03 May 13:37:14.939 # Next failover delay: I will not start a failover before Fri May  3 13:43:14 2019
40:X 03 May 13:43:14.808 # +new-epoch 4
40:X 03 May 13:43:14.816 # +vote-for-leader 5549d29a0710c611c7ea55c95d51085ca5d8913c 4
40:X 03 May 13:43:14.820 # Next failover delay: I will not start a failover before Fri May  3 13:49:15 2019
40:X 03 May 13:49:15.632 # +new-epoch 5
40:X 03 May 13:49:15.632 # +try-failover master mymaster 172.17.0.4 6379
40:X 03 May 13:49:15.647 # +vote-for-leader 71c9288565806da082b428463d07735bd08a9427 5
40:X 03 May 13:49:15.665 # 5549d29a0710c611c7ea55c95d51085ca5d8913c voted for 71c9288565806da082b428463d07735bd08a9427 5
40:X 03 May 13:49:15.731 # +elected-leader master mymaster 172.17.0.4 6379
40:X 03 May 13:49:15.731 # +failover-state-select-slave master mymaster 172.17.0.4 6379
40:X 03 May 13:49:15.814 # -failover-abort-no-good-slave master mymaster 172.17.0.4 6379
40:X 03 May 13:49:15.898 # Next failover delay: I will not start a failover before Fri May  3 13:55:16 2019
40:X 03 May 13:55:16.184 # +new-epoch 6
40:X 03 May 13:55:16.184 # +try-failover master mymaster 172.17.0.4 6379
40:X 03 May 13:55:16.194 # +vote-for-leader 71c9288565806da082b428463d07735bd08a9427 6
40:X 03 May 13:55:16.195 # 5549d29a0710c611c7ea55c95d51085ca5d8913c voted for 5549d29a0710c611c7ea55c95d51085ca5d8913c 6
40:X 03 May 13:55:26.819 # -failover-abort-not-elected master mymaster 172.17.0.4 6379
40:X 03 May 13:55:26.891 # Next failover delay: I will not start a failover before Fri May  3 14:01:16 2019
40:X 03 May 14:01:16.566 # +new-epoch 7
40:X 03 May 14:01:16.574 # +vote-for-leader 5549d29a0710c611c7ea55c95d51085ca5d8913c 7
40:X 03 May 14:01:16.634 # Next failover delay: I will not start a failover before Fri May  3 14:07:17 2019
40:X 03 May 14:07:16.880 # +new-epoch 8
40:X 03 May 14:07:16.888 # +vote-for-leader 5549d29a0710c611c7ea55c95d51085ca5d8913c 8
40:X 03 May 14:07:16.911 # Next failover delay: I will not start a failover before Fri May  3 14:13:16 2019
40:X 03 May 14:13:16.964 # +new-epoch 9
40:X 03 May 14:13:16.965 # +try-failover master mymaster 172.17.0.4 6379
40:X 03 May 14:13:16.975 # +vote-for-leader 71c9288565806da082b428463d07735bd08a9427 9
40:X 03 May 14:13:16.996 # 5549d29a0710c611c7ea55c95d51085ca5d8913c voted for 71c9288565806da082b428463d07735bd08a9427 9
40:X 03 May 14:13:17.034 # +elected-leader master mymaster 172.17.0.4 6379
40:X 03 May 14:13:17.034 # +failover-state-select-slave master mymaster 172.17.0.4 6379
40:X 03 May 14:13:17.118 # -failover-abort-no-good-slave master mymaster 172.17.0.4 6379
40:X 03 May 14:13:17.202 # Next failover delay: I will not start a failover before Fri May  3 14:19:17 2019
40:X 03 May 14:19:17.084 # +new-epoch 10
40:X 03 May 14:19:17.084 # +try-failover master mymaster 172.17.0.4 6379
40:X 03 May 14:19:17.094 # +vote-for-leader 71c9288565806da082b428463d07735bd08a9427 10
40:X 03 May 14:19:17.111 # 5549d29a0710c611c7ea55c95d51085ca5d8913c voted for 71c9288565806da082b428463d07735bd08a9427 10
40:X 03 May 14:19:17.195 # +elected-leader master mymaster 172.17.0.4 6379
40:X 03 May 14:19:17.195 # +failover-state-select-slave master mymaster 172.17.0.4 6379
40:X 03 May 14:19:17.296 # -failover-abort-no-good-slave master mymaster 172.17.0.4 6379
40:X 03 May 14:19:17.352 # Next failover delay: I will not start a failover before Fri May  3 14:25:17 2019
40:X 03 May 14:25:17.222 # +new-epoch 11
40:X 03 May 14:25:17.222 # +try-failover master mymaster 172.17.0.4 6379
40:X 03 May 14:25:17.233 # +vote-for-leader 71c9288565806da082b428463d07735bd08a9427 11
40:X 03 May 14:25:17.251 # 5549d29a0710c611c7ea55c95d51085ca5d8913c voted for 71c9288565806da082b428463d07735bd08a9427 11
40:X 03 May 14:25:17.323 # +elected-leader master mymaster 172.17.0.4 6379
40:X 03 May 14:25:17.324 # +failover-state-select-slave master mymaster 172.17.0.4 6379
40:X 03 May 14:25:17.391 # -failover-abort-no-good-slave master mymaster 172.17.0.4 6379
40:X 03 May 14:25:17.453 # Next failover delay: I will not start a failover before Fri May  3 14:31:18 2019
40:X 03 May 14:31:18.175 # +new-epoch 12
40:X 03 May 14:31:18.175 # +try-failover master mymaster 172.17.0.4 6379
40:X 03 May 14:31:18.183 # +vote-for-leader 71c9288565806da082b428463d07735bd08a9427 12
40:X 03 May 14:31:18.193 # 5549d29a0710c611c7ea55c95d51085ca5d8913c voted for 71c9288565806da082b428463d07735bd08a9427 12
40:X 03 May 14:31:18.267 # +elected-leader master mymaster 172.17.0.4 6379
40:X 03 May 14:31:18.267 # +failover-state-select-slave master mymaster 172.17.0.4 6379
40:X 03 May 14:31:18.335 # -failover-abort-no-good-slave master mymaster 172.17.0.4 6379
40:X 03 May 14:31:18.391 # Next failover delay: I will not start a failover before Fri May  3 14:37:19 2019
40:X 03 May 14:37:18.759 # +new-epoch 13
40:X 03 May 14:37:18.768 # +vote-for-leader 5549d29a0710c611c7ea55c95d51085ca5d8913c 13
40:X 03 May 14:37:18.787 # Next failover delay: I will not start a failover before Fri May  3 14:43:19 2019
40:X 03 May 14:43:18.927 # +new-epoch 14
40:X 03 May 14:43:18.935 # +vote-for-leader 5549d29a0710c611c7ea55c95d51085ca5d8913c 14
40:X 03 May 14:43:18.959 # Next failover delay: I will not start a failover before Fri May  3 14:49:19 2019
40:X 03 May 14:49:19.857 # +new-epoch 15
40:X 03 May 14:49:19.857 # +try-failover master mymaster 172.17.0.4 6379
40:X 03 May 14:49:19.865 # +vote-for-leader 71c9288565806da082b428463d07735bd08a9427 15
40:X 03 May 14:49:19.883 # 5549d29a0710c611c7ea55c95d51085ca5d8913c voted for 71c9288565806da082b428463d07735bd08a9427 15
40:X 03 May 14:49:19.949 # +elected-leader master mymaster 172.17.0.4 6379
40:X 03 May 14:49:19.949 # +failover-state-select-slave master mymaster 172.17.0.4 6379
40:X 03 May 14:49:20.021 # -failover-abort-no-good-slave master mymaster 172.17.0.4 6379
40:X 03 May 14:49:20.094 # Next failover delay: I will not start a failover before Fri May  3 14:55:20 2019
```

File: tasks/setup_redis.yml [hunk 6]
```diff
@@ -0,0 +1,6 @@
+---
+- name: disable redis binding to ::1(ipv6)
+  replace:
+    path: /etc/redis/redis.conf
+    regexp: 'bind 127.0.0.1.*'
+    replace: 'bind 0.0.0.0'
```
Comment (kvaps) [inline tasks/setup_redis.yml]: There is required
```
systemctl restart redis-server
```
to apply this change

**Pull request closed without merging**